gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java

```diff
@@ -132,29 +149,113 @@ public static void renamePath(FileSystem fs, Path oldName, Path newName) throws
+   * @param deleteSource true if the source should be deleted after the copy is complete, false otherwise
+   * @param overwriteDst true if the destination should be overwritten, false otherwise
+   * @param overwriteTmp true if the temp file should be overwritten, false otherwise
    */
-  public static void movePath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst) throws IOException {
-    if (srcFs.getUri().equals(dstFs.getUri())) {
-      renamePath(srcFs, src, dst);
+  public static void copyPath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst, Path tmp, boolean deleteSource,
+      boolean overwriteDst, boolean overwriteTmp, Configuration conf)
+      throws IOException {
+
+    if (S3_SCHEMES.contains(srcFs.getUri().getScheme()) || S3_SCHEMES.contains(dstFs.getUri().getScheme())) {
+      copyPath(srcFs, src, dstFs, dst, deleteSource, overwriteDst, conf);
     } else {
-      if (!FileUtil.copy(srcFs, src, dstFs, dst, true, false, dstFs.getConf())) {
-        throw new IOException(String.format("Failed to move %s to %s", src, dst));
+      if (!overwriteDst && dstFs.exists(dst)) {
+        throw new IOException(String.format("Cannot copy %s to %s because the dst exists", src, dst));
+      }
+      copyPath(srcFs, src, dstFs, tmp, false, overwriteTmp, conf);
+      try {
+        if (overwriteDst && dstFs.exists(dst)) {
+          deletePath(dstFs, dst, true);
+        }
+        renamePath(dstFs, tmp, dst);
+      } finally {
+        deletePath(dstFs, tmp, false);
+      }
+      if (deleteSource) {
+        srcFs.delete(src, false);
+      }
+    }
+  }
+
+  /**
+   * Copies data from a src {@link Path} to a dst {@link Path}.
+   *
+   * <p>
+   *   This method should be used in preference to
+   *   {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, boolean, Configuration)}, which does not handle
+   *   clean up of incomplete files if there is an error while copying data.
+   * </p>
+   *
+   * <p>
+   *   Generally, it is better to use
+   *   {@link #copyPath(FileSystem, Path, FileSystem, Path, Path, boolean, boolean, boolean, Configuration)} over this
+   *   method. The aformentioned method firsts writes data to a tmp file before committing it, which is usually
+   *   preferable to copying data directly to the destination.
+   * </p>
+   *
+   * @param srcFs the source {@link FileSystem} where the src {@link Path} exists
+   * @param src the {@link Path} to copy from the source {@link FileSystem}
+   * @param dstFs the destination {@link FileSystem} where the dst {@link Path} should be created
+   * @param dst the {@link Path} to copy data to
+   * @param deleteSource true if the source should be deleted after the copy is complete, false otherwise
+   * @param overwriteDst true if the destination should be overwritten, false otherwise
+   */
+  public static void copyPath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst, boolean deleteSource,
```

Comment:
As stated above, I really think that this should be transparent for users/developers. We should default to the safest, but possibly less performant, methods of performing actions.  In the case of filesystems, we should be able to decorate specific filesystems with more performant versions.  For instance, the base copy method could be the local->temp->dest version.  But, the s3* filesystems could be decorated to perform local->dest.